crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3054,7 +3069,20 @@ impl<'db> TypeInferenceBuilder<'db> {
             ctx: _,
         } = attribute;
 
+        // If value is a module, and we have imported a submodule of that module named [attr], then
+        // the result is that submodule, even if the module also defines a (non-module) symbol with
+        // that name.
         let value_ty = self.infer_expression(value);
+        if let Type::ModuleLiteral(module) = &value_ty {
+            let mut submodule_name = module.name(self.db).clone();
+            submodule_name.append(&attr.id);
+            if self.index.imports_module(&submodule_name) {
+                if let Some(submodule_ty) = self.module_ty_from_name(&submodule_name) {
+                    return submodule_ty;
+                }
+            }
+        }
```

Comment:
Nit: Could this logic be moved into `Type::member` (which then calls into `ModuleType::member`) or does it only apply when looking up attributes?

Reply:
I don't think it can be in `Type` because it requires resolving the module, which is currently handled by `TypeInferenceBuilder`.  And the "submodules are available as attributes of the containing package" bit is only relevant during attribute access.

Reply:
That said, resolving the module doesn't look like it _needs_ to be in `TypeInferenceBuilder`, since it's just calling `resolve_module`.  Let me see if I can move it into `Type` after all

Reply:
Sorry, no, scratch that — it needs access to the semantic index of the containing file to see which modules have been imported.  So the two options are:

- Leave it here
- Add the containing `File` to `ModuleLiteralType` so that we can move this logic into `Type::member`

Reply:
Latest patch adds the containing file to `ModuleLiteralType`.  Interestingly, that allowed one of the expected-failure mdtests to [start succeeding](https://github.com/astral-sh/ruff/pull/14946/commits/5fb1246cbe4dac1529defb990a8fb5c12c00b6a0#diff-5c8508e2bc14c029e901b266c723a52bbf1b574be8129942162b108492b77fcf)!